narwhal/config/src/committee.rs

```diff
@@ -0,0 +1,512 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{CommitteeUpdateError, ConfigError, Epoch, Stake};
+use crypto::{NetworkPublicKey, PublicKey, PublicKeyBytes};
+use fastcrypto::traits::EncodeDecodeBase64;
+use mysten_network::Multiaddr;
+use mysten_util_mem::MallocSizeOf;
+use once_cell::sync::OnceCell;
+use rand::rngs::StdRng;
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashSet};
+use std::fmt::{Display, Formatter};
+
+#[derive(Clone, Serialize, Deserialize, Debug, Eq, PartialEq)]
+pub struct Authority {
+    /// The id under which we identify this authority across Narwhal
+    #[serde(skip)]
+    id: OnceCell<AuthorityIdentifier>,
+    /// The authority's main PublicKey which is used to verify the content they sign.
+    #[serde(skip)]
+    protocol_key: OnceCell<PublicKey>,
+    /// The authority's main PublicKey expressed as pure bytes
+    #[serde(skip)]
+    protocol_key_bytes: OnceCell<PublicKeyBytes>,
+    /// The voting power of this authority.
+    stake: Stake,
+    /// The network address of the primary.
+    primary_address: Multiaddr,
+    /// Network key of the primary.
+    network_key: NetworkPublicKey,
+}
+
+impl Authority {
+    /// The constructor is not public by design. Everyone who wants to create authorities should do
+    /// it via Committee (more specifically can use CommitteeBuilder). As some internal properties of
+    /// Authority are initialised via the Committee, to ensure that the user will not accidentally use
+    /// stale Authority data, should always derive them via the Commitee.
+    fn new(
+        protocol_key: PublicKey,
+        stake: Stake,
+        primary_address: Multiaddr,
+        network_key: NetworkPublicKey,
+    ) -> Self {
+        let protocol_key_bytes = PublicKeyBytes::from(&protocol_key);
+
+        Self {
+            id: Default::default(),
+            protocol_key: OnceCell::with_value(protocol_key),
```

Comment:
Do we really need OnceCell for protocol_key and protocol_key_bytes?

Reply:
For similar reason as above I've done this to avoid issues when initialising via a file - which we can totally avoid by changing the logic there - I've done it like this for the time being to not break things , but I am willing to iterate and simplify

Reply:
I've removed `OnceCell` - I thought about it again and the file read as a reason on its own is not enough to get the extra overhead